### DISCUSSION TITLE
How can I create a user on login and proceed with login automatically?

### DISCUSSION THREAD

**Discussion by tylerforesthauser** [directus/directus#12988] [Archive]
Using external data from a trusted source, I want to be able to essentially construct a user and transparently log it in when a user logs in for the first time. I've tried approaching this a couple ways with varied success.

1. I've built a custom login endpoint that more or less duplicates the logic of the default endpoint but shims in my user lookup/building logic. This works but is less than ideal (I ended up creating matching login, refresh and user endpoints just for consistency). 

2. The other route I've tried is using the auth.login filter hook + pre-importing my userdata but storing the password as NULL. This allows me to reach the auth.login hook, use the login payload to confirm that my user can successfully log in, and then update my Directus db with the user password via UserService. This makes more sense in terms of code maintainability and is less confusing, however, I've hit a roadblock with this approach, as I'll detail below.

I've found that inside the Authentication Service, the user is being [queried](https://github.com/directus/directus/blob/main/api/src/services/authentication.ts#L52-L72) _prior_ to the [filter event emitter](https://github.com/directus/directus/blob/main/api/src/services/authentication.ts#L74-L87) and then it's that original user data being used to perform the [actual login](https://github.com/directus/directus/blob/main/api/src/services/authentication.ts#L142), rendering any changes made to the user useless until the next time the filter event is emitted again. This means that the user essentially needs to try logging in twice to successfully log in.

I'm hoping perhaps someone has some clever way to get around these circumstances to accomplish what I'm aiming to do. Is there any way I can "jump the gun", so to speak, and get the user logged in as part of the auth.login filter event, perhaps? Or is there some other way I could be approaching this?

**Comment by rijkvanzanten** [comment]
I think we'll have to think through a new before-login-attempt hook for this sort of use case 🤔

**Comment by tylerforesthauser** [comment]
So my "super clever" way of getting around this limitation is as follows:

I've created a temporary password that is assigned to all imported users (defined in my .env file) and I've added an additional field flagging the user as not yet activated. Checking to see that this flag is set _and_ using the AuthenticationService's `verifyPassword` method to confirm the user's password is set to the temporary password, I proceed to validate the user's provided login credentials with my external source, update the database user with the validated user password (and unset the activation flag), and then set the payload password to the temporary password and return the payload. This allows the initial login to occur and every subsequent login will operate normally as the activation flag will not be set and the password won't match the temporary password.

This should suffice for now but I would _definitely_ love to see some kind of "before-login-attempt hook" added in the future.
